pkg/builder/controller.go

```diff
@@ -259,12 +255,12 @@ func (blder *Builder) doWatch() error {
 		allPredicates = append(allPredicates, w.predicates...)
 
 		// If the source of this watch is of type *source.Kind, project it.
```

Comment:
nit: Comment seems to be outdated

Reply:
Q: is it correct to check for SyncingSource and then pass the src into KindAsPartialMetadata where we check for kind?

Could it be a SyncingSource which is not kind? ("Users may build their own Source implementations.")

Reply:
I agree it's not ideal. Originally I had a `KindSource` instead with `AsPartialMetadata()` method required. These builders need a lot of rework honestly, the UX also isn't ideal with modifiers in place. I'm happy to go back to the interface if we prefer that.

Reply:
PTAL, added a new interface called `TypedSource` which lets users provide their own if needed

Reply:
lgtm